### PR TITLE
Allow inplacing buffer when other users are inconsequential

### PR DESCRIPTION
**Pull request by exclamaforte**
Summary:
I think we can inplace a buffer if all of the users of said buffer are "inconsequential", defined as having been removed, being completed, or being part of the ancestors set. In particular, this allows LayerNorm to inplace its input buffer.

Implements:
https://github.com/pytorch/pytorch/issues/132826

Test Plan:
New unit test of matmul followed by LayerNorm, make sure there's an inplaced buffer.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/scheduler.py
Comment (exclamaforte) [inline torch/_inductor/scheduler.py:412]: I'm not sure if self can be in ancestors at this point in the IR, open to feedback from reviews.

File: torch/_inductor/scheduler.py
Comment (eellison) [inline torch/_inductor/scheduler.py:413]: i'm surprised a node is an ancestor of itself...

File: test/inductor/test_torchinductor.py
Comment (eellison) [inline test/inductor/test_torchinductor.py:12548]: nit - can make test work for both devices by passing in self.device and using above

**Review by eellison — APPROVED** [review]
Cool! nice to see this required minimal changes. cc @shunting314 

Before we land - maybe let's do a dashboard run for safety. this should be motivation for landing donated buffers and  turning on https://github.com/pytorch/pytorch/pull/133368 cc @BoyuanFeng

**Review by shunting314 — COMMENTED** [review]
Cool!

Would you share the generated in-place kernel?

Can we do some benchmarking on the toy examples regarding
1. memory saving
2. perf impact

File: torch/_inductor/scheduler.py
Comment (exclamaforte) [reply torch/_inductor/scheduler.py:413]: I'm not sure if it's possible, do you know if the IR has been de-cycled when it gets to the scheduler? I added it defensibly in case there was a cycle, to prevent inplacing.

Comment (eellison) [reply torch/_inductor/scheduler.py:413]: Pretty sure a node should not be an ancestor of itself - you can add an assertion, run tests, then remove

**Comment by exclamaforte** [comment]
Looks like the perf impact is slightly negative 🤔 
Screenshot 2024-10-28 at 10 59 57 AM

**Comment by eellison** [comment]
Is the base commit of your benchmarking run the same one you are comparing to ?

**Comment by exclamaforte** [comment]
Yeah unfortunately
Screenshot 2024-10-28 at 11 14 35 AM

**Comment by exclamaforte** [comment]
Ah cool! Sorry, still getting used to the perf dashboard, so I got the order flipped. Anyways, good to merge on perf after tests pass?

**Comment by shunting314** [comment]
Can you share a generated inplace kernel? In some odd situations, Inductor codegen an 'in_out_ptr' but does not really do load/store..

**Comment by eellison** [comment]
@shunting314 i think that was previously fixed: https://github.com/pytorch/pytorch/issues/120217

@exclamaforte should be good to land, yes

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but it seems to break trunk after landing' -c nosignal

Here is the failed test, but I'm not entirely sure your PR is the cause.  So I will revert it to see if trunk recovers, and will reland the change if it doesn't.

inductor/test_cooperative_reductions.py::CooperativeReductionTests::test_chained_reductions [GH job link](https://github.com/pytorch/pytorch/actions/runs/11602039117/job/32307770322) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/8840889c3f6565b7975150adebcbe062f19035ee)

**Comment by eellison** [comment]
Just came across another case where this would help - would be great to get this landed

**Comment by exclamaforte** [comment]
@eellison absolutely I'm just making sure it doesn't break cooperative reductions:
https://github.com/pytorch/pytorch/pull/137756

**Comment by exclamaforte** [comment]
The only interaction is just removing some of the buffers from the kernel with inplacing, so I think it's fine. Here is the diff in the generated kernels:
```
--- original-sc.py	2024-10-31 13:21:53.102671430 -0700
+++ my-change-sc.py	2024-10-31 13:18:58.715538419 -0700
@@ -37,7 +37,7 @@
 async_compile = AsyncCompile()
 
 
-# kernel path: /tmp/tmpfqaxv1a3/bb/cbbvu2avvkew2z2exvvc3miqd74gcqrn56mibaj5d5mx2ar6ijyl.py
+# kernel path: /tmp/tmpecuu9ih2/qa/cqa7bdgfux4ejjav4saapta6ij7mr3xkayff5jfxm4devpjloavy.py
 # Topologically Sorted Source Nodes: [softmax, x, softmax_1, x_1, softmax_2, x_2, softmax_3, x_3, softmax_4, x_4, softmax_5, x_5, softmax_6, x_6, softmax_7, x_7], Original ATen: [aten._softmax, aten.add]
 # Source node to ATen node mapping:
 #   softmax => amax, div, exp, sub, sum_1
@@ -119,11 +119,11 @@
     size_hints=[4, 131072],
     reduction_hint=ReductionHint.INNER,
     filename=__file__,
-    triton_meta={'signature': {'in_ptr0': '*fp32', 'out_ptr4': '*fp32', 'out_ptr9': '*fp32', 'out_ptr14': '*fp32', 'out_ptr19': '*fp32', 'sem_ptr': '*u32', 'ws_ptr': '*u8', 'xnumel': 'i32', 'rnumel': 'i32'}, 'device': DeviceProperties(type='cuda', index=0, cc=90, major=9, regs_per_multiprocessor=65536, max_threads_per_multi_processor=2048, multi_processor_count=132, warp_size=32), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2, 3, 4, 5, 6, 8), equal_to_1=())]},
-    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_unk_fused__softmax_add_0', 'mutated_arg_names': [], 'optimize_mem': True, 'no_x_dim': False, 'num_load': 17, 'num_reduction': 16, 'backend_hash': '56D3B085EC0EDBE6B2FD2331DF13B46A9D62F65F48BC9B3C8C1106221DAF119E', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'persistent_reduction': False}
+    triton_meta={'signature': {'in_out_ptr0': '*fp32', 'in_ptr0': '*fp32', 'sem_ptr': '*u32', 'ws_ptr': '*u8', 'xnumel': 'i32', 'rnumel': 'i32'}, 'device': DeviceProperties(type='cuda', index=0, cc=90, major=9, regs_per_multiprocessor=65536, max_threads_per_multi_processor=2048, multi_processor_count=132, warp_size=32), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2, 3, 5), equal_to_1=())]},
+    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_unk_fused__softmax_add_0', 'mutated_arg_names': ['in_out_ptr0'], 'optimize_mem': True, 'no_x_dim': False, 'num_load': 17, 'num_reduction': 16, 'backend_hash': '56D3B085EC0EDBE6B2FD2331DF13B46A9D62F65F48BC9B3C8C1106221DAF119E', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'persistent_reduction': False}
 )
 @triton.jit
-def triton_unk_fused__softmax_add_0(in_ptr0, out_ptr4, out_ptr9, out_ptr14, out_ptr19, sem_ptr, ws_ptr, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr, RSPLIT : tl.constexpr):
+def triton_unk_fused__softmax_add_0(in_out_ptr0, in_ptr0, sem_ptr, ws_ptr, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr, RSPLIT : tl.constexpr):
     xnumel = 4
     rnumel = 100000
     rsplit_id = tl.program_id(0)
@@ -238,7 +238,7 @@
         tmp37 = tl.broadcast_to(tmp36, [XBLOCK, RBLOCK])
         tmp39 = triton_helpers.maximum(_tmp38, tmp37)
         _tmp38 = tl.where(rmask & xmask, tmp39, _tmp38)
-        tl.store(out_ptr4 + (r1 + (100000*x0)), tmp36, rmask & xmask)
+        tl.store(in_out_ptr0 + (r1 + (100000*x0)), tmp36, rmask & xmask)
     tmp38 = triton_helpers.max2(_tmp38, 1)[:, None]
     if RSPLIT > 1:
         tmp38_ws = (ws_ptr + 2048).to(tl.pointer_type(tl.float32))
@@ -253,7 +253,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp40 = tl.load(out_ptr4 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp40 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp41 = tmp40 - tmp38
         tmp42 = tl_math.exp(tmp41)
         tmp43 = tl.broadcast_to(tmp42, [XBLOCK, RBLOCK])
@@ -273,7 +273,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp46 = tl.load(out_ptr4 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp46 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp47 = tmp46 - tmp38
         tmp48 = tl_math.exp(tmp47)
         tmp49 = tmp48 / tmp44
@@ -295,7 +295,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp54 = tl.load(out_ptr4 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp54 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp55 = tmp54 - tmp38
         tmp56 = tl_math.exp(tmp55)
         tmp57 = tmp56 / tmp44
@@ -319,7 +319,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp64 = tl.load(out_ptr4 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_first', other=0.0)
+        tmp64 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp65 = tmp64 - tmp38
         tmp66 = tl_math.exp(tmp65)
         tmp67 = tmp66 / tmp44
@@ -331,7 +331,7 @@
         tmp73 = tl.broadcast_to(tmp72, [XBLOCK, RBLOCK])
         tmp75 = triton_helpers.maximum(_tmp74, tmp73)
         _tmp74 = tl.where(rmask & xmask, tmp75, _tmp74)
-        tl.store(out_ptr9 + (r1 + (100000*x0)), tmp72, rmask & xmask)
+        tl.store(in_out_ptr0 + (r1 + (100000*x0)), tmp72, rmask & xmask)
     tmp74 = triton_helpers.max2(_tmp74, 1)[:, None]
     if RSPLIT > 1:
         tmp74_ws = (ws_ptr + 4096).to(tl.pointer_type(tl.float32))
@@ -346,7 +346,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp76 = tl.load(out_ptr9 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp76 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp77 = tmp76 - tmp74
         tmp78 = tl_math.exp(tmp77)
         tmp79 = tl.broadcast_to(tmp78, [XBLOCK, RBLOCK])
@@ -366,7 +366,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp82 = tl.load(out_ptr9 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp82 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp83 = tmp82 - tmp74
         tmp84 = tl_math.exp(tmp83)
         tmp85 = tmp84 / tmp80
@@ -388,7 +388,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp90 = tl.load(out_ptr9 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp90 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp91 = tmp90 - tmp74
         tmp92 = tl_math.exp(tmp91)
         tmp93 = tmp92 / tmp80
@@ -412,7 +412,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp100 = tl.load(out_ptr9 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_first', other=0.0)
+        tmp100 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp101 = tmp100 - tmp74
         tmp102 = tl_math.exp(tmp101)
         tmp103 = tmp102 / tmp80
@@ -424,7 +424,7 @@
         tmp109 = tl.broadcast_to(tmp108, [XBLOCK, RBLOCK])
         tmp111 = triton_helpers.maximum(_tmp110, tmp109)
         _tmp110 = tl.where(rmask & xmask, tmp111, _tmp110)
-        tl.store(out_ptr14 + (r1 + (100000*x0)), tmp108, rmask & xmask)
+        tl.store(in_out_ptr0 + (r1 + (100000*x0)), tmp108, rmask & xmask)
     tmp110 = triton_helpers.max2(_tmp110, 1)[:, None]
     if RSPLIT > 1:
         tmp110_ws = (ws_ptr + 6144).to(tl.pointer_type(tl.float32))
@@ -439,7 +439,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp112 = tl.load(out_ptr14 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp112 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp113 = tmp112 - tmp110
         tmp114 = tl_math.exp(tmp113)
         tmp115 = tl.broadcast_to(tmp114, [XBLOCK, RBLOCK])
@@ -459,7 +459,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp118 = tl.load(out_ptr14 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp118 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp119 = tmp118 - tmp110
         tmp120 = tl_math.exp(tmp119)
         tmp121 = tmp120 / tmp116
@@ -481,7 +481,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp126 = tl.load(out_ptr14 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
+        tmp126 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_last', other=0.0)
         tmp127 = tmp126 - tmp110
         tmp128 = tl_math.exp(tmp127)
         tmp129 = tmp128 / tmp116
@@ -504,7 +504,7 @@
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
-        tmp136 = tl.load(out_ptr14 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_first', other=0.0)
+        tmp136 = tl.load(in_out_ptr0 + (r1 + (100000*x0)), rmask & xmask, eviction_policy='evict_first', other=0.0)
         tmp137 = tmp136 - tmp110
         tmp138 = tl_math.exp(tmp137)
         tmp139 = tmp138 / tmp116
@@ -513,7 +513,7 @@
         tmp142 = tl_math.exp(tmp141)
         tmp143 = tmp142 / tmp134
         tmp144 = tmp140 + tmp143
-        tl.store(out_ptr19 + (r1 + (100000*x0)), tmp144, rmask & xmask)
+        tl.store(in_out_ptr0 + (r1 + (100000*x0)), tmp144, rmask & xmask)
 ''', device_str='cuda')
 
 
@@ -527,20 +527,17 @@
     with torch.cuda._DeviceGuard(0):
         torch.cuda.set_device(0)
         buf4 = empty_strided_cuda((4, 100000), (100000, 1), torch.float32)
-        buf9 = empty_strided_cuda((4, 100000), (100000, 1), torch.float32)
-        buf14 = empty_strided_cuda((4, 100000), (100000, 1), torch.float32)
-        buf19 = empty_strided_cuda((4, 100000), (100000, 1), torch.float32)
+        buf9 = buf4; del buf4  # reuse
+        buf14 = buf9; del buf9  # reuse
+        buf19 = buf14; del buf14  # reuse
         # Topologically Sorted Source Nodes: [softmax, x, softmax_1, x_1, softmax_2, x_2, softmax_3, x_3, softmax_4, x_4, softmax_5, x_5, softmax_6, x_6, softmax_7, x_7], Original ATen: [aten._softmax, aten.add]
         semaphores_cuda_0 = empty_strided_cuda((4, ), (1, ), torch.uint32)
         semaphores_cuda_0.zero_()
         workspace_0 = empty_strided_cuda((8192, ), (1, ), torch.uint8)
         stream0 = get_raw_stream(0)
-        triton_unk_fused__softmax_add_0.run(arg0_1, buf4, buf9, buf14, buf19, semaphores_cuda_0, workspace_0, 4, 100000, grid=cooperative_reduction_grid(4), stream=stream0)
+        triton_unk_fused__softmax_add_0.run(buf19, arg0_1, semaphores_cuda_0, workspace_0, 4, 100000, grid=cooperative_reduction_grid(4), stream=stream0)
         del workspace_0
         del arg0_1
-        del buf14
-        del buf4
-        del buf9
     return (buf19, )
```

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting this again, but I think it has a test failing internally and also on ROCm'

Internal `buck2 test '@fbcode//mode/opt' fbcode//caffe2/test/inductor:benchmark_fusion -- --exact 'caffe2/test/inductor:benchmark_fusion - test_avoid_register_spilling_cuda (caffe2.test.inductor.test_benchmark_fusion.BenchmarkFusionCudaTest)'`

ROCm OSS:
inductor/test_benchmark_fusion.py::BenchmarkFusionCudaTest::test_avoid_register_spilling_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/11639450838/job/32416092379) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/540f3ef9b1873923dde420453ab69fc9cf67fb69)

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting this again, but I think it has a test failing internally and also on ROCm' -c ghfirst

Internal D65342635 `buck2 test '@fbcode//mode/opt' fbcode//caffe2/test/inductor:benchmark_fusion -- --exact 'caffe2/test/inductor:benchmark_fusion - test_avoid_register_spilling_cuda (caffe2.test.inductor.test_benchmark_fusion.BenchmarkFusionCudaTest)'`

ROCm OSS:
inductor/test_benchmark_fusion.py::BenchmarkFusionCudaTest::test_avoid_register_spilling_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/11639450838/job/32416092379) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/540f3ef9b1873923dde420453ab69fc9cf67fb69)

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting this again, but I think it has a test failing internally and also on ROCm' -c ghfirst

Internal D65342635 `buck2 test '@fbcode//mode/opt' fbcode//caffe2/test/inductor:benchmark_fusion -- --exact 'caffe2/test/inductor:benchmark_fusion - test_avoid_register_spilling_cuda (caffe2.test.inductor.test_benchmark_fusion.BenchmarkFusionCudaTest)'`

ROCm OSS:
inductor/test_benchmark_fusion.py::BenchmarkFusionCudaTest::test_avoid_register_spilling_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/11639450838/job/32416092379) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/540f3ef9b1873923dde420453ab69fc9cf67fb69)

**Pull request closed without merging**
